kasa/protocol.py

```diff
@@ -126,3 +134,152 @@ def decrypt(ciphertext: bytes) -> str:
         plaintext = bytes(buffer)
 
         return plaintext.decode()
+
+
+class TPLinkKLAP:
+    """Implementation of the KLAP encryption protocol.
+
+    KLAP is the name used in device discovery for TP-Link's new encryption
+    protocol, which appeared with firmware 1.1.0.
+    """
+
+    def __init__(self, host: str, authentication) -> None:
+        self.host = host
+        self.jar = aiohttp.CookieJar(unsafe=True, quote_cookie=False)
+        self.clientBytes = secrets.token_bytes(16)
+        self.authenticator = self.__computeAuthenticator(authentication)
+        self.handshake_lock = asyncio.Lock()
+        self.handshake_done = False
+
+        _LOGGER.debug("[KLAP] Created KLAP object for %s", self.host)
+
+    @staticmethod
+    def __computeAuthenticator(authentication: Optional[dict]) -> bytes:
+        if authentication is not None:
+            username = authentication["user"].encode()
+            password = authentication["password"].encode()
+        else:
+            username = bytearray(b"")
+            password = bytearray(b"")
+        _LOGGER.debug("[KLAP] Using username %s and password %s", username, password)
+        return hashlib.md5(
+            hashlib.md5(username).digest() + hashlib.md5(password).digest()
+        ).digest()
```

Comment:
See my comment above on hashlib repetition. Maybe even create a helper function (`md5digest()`) to avoid repetition?

Reply:
Done - this code has also moved to the new Auth class.